### PR TITLE
Keras tuner example.

### PR DESCRIPTION
**Pull request by pavithrasv**
Simple keras tuner example.

### CODE REVIEW

File: tensorflow_cloud/machine_config.py [hunk 1]
```diff
@@ -151,12 +151,6 @@ def validate(self):
         accelerator_type=AcceleratorType.NVIDIA_TESLA_V100,
         accelerator_count=4,
     ),
-    "V100_8X": MachineConfig(
```
Comment (pavithrasv) [inline tensorflow_cloud/machine_config.py:154]: Removed this as 8 V100s cannot be created. It is above the allowed limit of 4 for Cloud ML engine

File: tests/testdata/keras_tuner_mnist_example.py [hunk 2]
```diff
@@ -0,0 +1,48 @@
+from tensorflow import keras
+from tensorflow.keras import layers
+
+
+from kerastuner.tuners import RandomSearch
+from kerastuner.engine.hypermodel import HyperModel
+from kerastuner.engine.hyperparameters import HyperParameters
+
+
+(x, y), (val_x, val_y) = keras.datasets.mnist.load_data()
```
Comment (fchollet) [inline tests/testdata/keras_tuner_mnist_example.py:10]: I would recommend using a CIFAR10 convnet, to justify the use of the V100_4X config. A state-of-the-art CIFAR10 convnet should be achievable via random search in a reasonable amount of time.

I would also recommend starting the model with image data augmentation layers and including optional dropout in the search (before the classification layer)

File: tests/integration/call_run_on_script_with_keras_tuner_search.py [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+
+# Automated OneDeviceStrategy: default 1 GPU chief config and no workers
+# tfc.run(
+#     entry_point='tests/testdata/keras_tuner_mnist_example.py',
+#     distribution_strategy='auto',
+#     requirements_txt='tests/testdata/requirements.txt',
+#     stream_logs=True)
+
+# Automated MirroredStrategy: chief config with multiple GPUs
+tfc.run(
```
Comment (fchollet) [inline tests/integration/call_run_on_script_with_keras_tuner_search.py:37]: With this config, where are the models / search results saved?

File: tests/testdata/keras_tuner_mnist_example.py [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+    max_trials=5,
+    executions_per_trial=3,
+    directory="test_dir",
+)
+
+tuner.search_space_summary()
+
+tuner.search(x=x, y=y, epochs=3, validation_data=(val_x, val_y))
+
+tuner.results_summary()
```
Comment (fchollet) [inline tests/testdata/keras_tuner_mnist_example.py:48]: Let's also include retrieving the best model after the search is run (where are they saved?)

File: tests/testdata/keras_tuner_mnist_example.py (hunk 2 above)
Comment (pavithrasv) [reply tests/testdata/keras_tuner_mnist_example.py]: Done.

File: tests/integration/call_run_on_script_with_keras_tuner_search.py (hunk 3 above)
Comment (pavithrasv) [reply tests/integration/call_run_on_script_with_keras_tuner_search.py:23]: Added a model save path.

File: tests/testdata/keras_tuner_mnist_example.py (hunk 4 above)
Comment (pavithrasv) [reply tests/testdata/keras_tuner_mnist_example.py]: Retrieving the best model and saving it in GCS path.

**Comment by pavithrasv** [comment]
Updated. Thank you Francois!

File: tests/testdata/keras_tuner_cifar_example.py [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+        loss="categorical_crossentropy",
+        optimizer=keras.optimizers.RMSprop(
+            learning_rate=hp.Choice("learning_rate", [1e-2, 1e-3, 1e-4]), decay=1e-6
+        ),
+        metrics=["accuracy"],
+    )
+    return model
+
+
+datagen = ImageDataGenerator(
```
Comment (fchollet) [inline tests/testdata/keras_tuner_cifar_example.py:79]: Using `ImageDataGenerator` in a single process setting will be slow, when using it you should also use `num_workers=N` and `use_multiprocessing=True`

Here I actually recommend using a dataset (from_tensor_slices) and starting the model with a preprocessing stage like this: https://keras.io/examples/vision/image_classification_from_scratch/#using-image-data-augmentation

File: tests/testdata/keras_tuner_cifar_example.py [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+y_train = keras.utils.to_categorical(y_train, 10)
+y_test = keras.utils.to_categorical(y_test, 10)
+x_train = x_train.astype("float32")
+x_test = x_test.astype("float32")
+x_train /= 255
+x_test /= 255
+
+
+def build_model(hp):
+    model = keras.Sequential()
```
Comment (fchollet) [inline tests/testdata/keras_tuner_cifar_example.py:24]: Comments about the architecture:

- No need for dropout after a Conv, only use dropout before a Dense
- To justify the use of 4 V100, we should do architecture search, not just search for the learning rate and dropout rates

I recommend the follow search:

```
activation = choice[ReLU, LeakyReLU, ELU]
for i in range(int[2-4]):  # blocks
  for j in range(int[1-3]):  # one block's content
    conv(int[16-256], same)
    activation
  maxpooling(2)  # end of block
dropout(float[0-1])
dense
```

File: tests/testdata/keras_tuner_cifar_example.py [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+    objective="val_accuracy",
+    max_trials=5,
+    executions_per_trial=3,
+    directory="test_dir",
+)
+
+tuner.search_space_summary()
+tuner.search(
+    datagen.flow(x_train, y_train, batch_size=32),
+    epochs=50,
```
Comment (fchollet) [inline tests/testdata/keras_tuner_cifar_example.py:98]: Increase the number of epochs (1000) and use EarlyStopping with ~3 epoch patience

File: tests/testdata/keras_tuner_cifar_example.py [hunk 8]
```diff
@@ -0,0 +1,112 @@
...
+                "dropout_3", min_value=0.0, max_value=0.5, default=0.25, step=0.05,
+            )
+        )
+    )
+    model.add(layers.Dense(10, activation="softmax"))
+
+    model.compile(
+        loss="categorical_crossentropy",
+        optimizer=keras.optimizers.RMSprop(
+            learning_rate=hp.Choice("learning_rate", [1e-2, 1e-3, 1e-4]), decay=1e-6
```
Comment (fchollet) [inline tests/testdata/keras_tuner_cifar_example.py:72]: Consider using a decay schedule for the learning rate, where you search both for the decay rate and the initial learning rate https://keras.io/api/optimizers/learning_rate_schedules/exponential_decay/

File: tests/testdata/keras_tuner_cifar_example.py (hunk 5 above)
Comment (pavithrasv) [reply tests/testdata/keras_tuner_cifar_example.py:79]: Done.

File: tests/testdata/keras_tuner_cifar_example.py (hunk 8 above)
Comment (pavithrasv) [reply tests/testdata/keras_tuner_cifar_example.py:72]: Done.

File: tests/testdata/keras_tuner_cifar_example.py (hunk 6 above)
Comment (pavithrasv) [reply tests/testdata/keras_tuner_cifar_example.py:24]: Done.

File: tests/testdata/keras_tuner_cifar_example.py (hunk 7 above)
Comment (pavithrasv) [reply tests/testdata/keras_tuner_cifar_example.py:98]: Done.

File: tests/testdata/keras_tuner_cifar_example.py (hunk 6 above)
Comment (pavithrasv) [reply tests/testdata/keras_tuner_cifar_example.py:24]: Used just 'relu' and 'elu' activations as hp.Choice does not allow objects of different types. I could not use layers.Activation('relu') and layers.LeakyRelu() together in hp.Choice.

**Comment by pavithrasv** [comment]
Updated, PTAL. Thank you Francois!
